numba/tests/test_interpreter.py

```diff
@@ -618,3 +618,202 @@ def inline_func(flag):
+    """
+
+    @skip_unless_py10
+    def test_large_heterogeneous_const_dict(self):
+        """
+        Tests that a function with a large heterogenous constant
+        dictionary remains a constant.
+        """
+        def const_func():
+            # D is a heterogenous dictionary
+            # so this code can only compile if
+            # d is constant.
+            d = {
+                "A": 1,
+                "B": 1,
+                "C": 1,
+                "D": 1,
+                "E": 1,
+                "F": 1,
+                "G": 1,
+                "H": 1,
+                "I": 1,
+                "J": 1,
+                "K": 1,
+                "L": 1,
+                "M": 1,
+                "N": 1,
+                "O": 1,
+                "P": 1,
+                "Q": 1,
+                "R": 1,
+                "S": 'a',
+            }
+            return d["S"]
+
+        py_func = const_func
+        cfunc = njit()(const_func)
+        a = py_func()
+        b = cfunc()
+        self.assertEqual(a, b)
+
+    @skip_unless_py10
+    def test_large_heterogeneous_const_keys_dict(self):
+        """
+        Tests that a function with a large heterogenous constant
+        dictionary remains a constant.
+        """
+        def const_keys_func(a):
+            # D is a heterogenous dictionary
+            # so this code can only compile if
+            # d has constant keys.
+            d = {
+                "A": 1,
+                "B": 1,
+                "C": 1,
+                "D": 1,
+                "E": 1,
+                "F": 1,
+                "G": 1,
+                "H": 1,
+                "I": 1,
+                "J": 1,
+                "K": 1,
+                "L": 1,
+                "M": 1,
+                "N": 1,
+                "O": 1,
+                "P": 1,
+                "Q": 1,
+                "R": 1,
+                "S": a,
+            }
+            return d["S"]
+
+        py_func = const_keys_func
+        cfunc = njit()(const_keys_func)
+        value = "efwf"
+        a = py_func(value)
+        b = cfunc(value)
+        self.assertEqual(a, b)
+
+    @skip_unless_py10
+    def test_usercode_update_heterogenous_keys(self):
+        """
+        Tests an example using update that mirrors
+        the pattern created by the Python 3.10 bytecode
+        changes. This verifies that additional code can
+        now be supported when applying this optimization.
+        """
+        def const_keys_func(a):
+            """
+            Dictionary update between two constant
+            dictionaries. When applying the optimization
+            it becomes possible to do the update at compile time.
+            """
+            d1 = {
+                "a": 1,
+                "b": 2,
+                "c": 3,
+            }
+            d2 = {
+                "d": 4,
+                "e": a
+            }
+            d1.update(d2)
+            return d1["e"]
```

Comment:
I agree this is potentially confusing. I added these tests because from my perspective this user code is the closest example that resembles the Python 3.10 bytecode changes (and thus cannot be distinguished in the current implementation). Writing code in this style also proved to be the easiest way to test edge cases not possible in just the bytecode changes (for example d2 being used after the update).

Personally I think this case is uncommon and its seems unlikely that people would write code like this, although I'm open to discussion about possible changes.

Reply:
I can understand implementing this from a "convenient way to test this specific transform" point of view, however, it has the unfortunate side effect of leaking into user facing functionality.

When designing and implementing the `LiteralStrKey` behaviour, it turned out there's a load of cases where you could potentially track some mutation of a const dictionary. However, from a usability perspective, it was decided that it'd be fairly hard to explain what's going when there's an invalid case even when the user actually wrote the control flow in their source (it being even harder when it occurs as a result of some bytecode optimisation!). All mutation of const types was therefore banned.

As a result of the above I do not think that supporting this should happen until:
1. a considerably improved method of tracking, explaining and reporting specific errors in relation to this is introduced.
2. all mutating methods in the type class are tracked and do the "mutation", this so as to provide consistency (it'd be strange to permit "update" but not "setitem").

Both of which I think are out of scope for this PR and would be a lot of work in themselves.